### DISCUSSION TITLE
Jamulus Directory Server vs quake Masterserver

### DISCUSSION THREAD

**Discussion by QJKX** [jamulussoftware/jamulus#2193] [Ideas]
There are some very obvious parallels between Jamulus and quake-like games, and the corresponding directory/masterserver.

https://333networks.com/howitworks

Is there anything Jamulus can learn or leverage off those existing masterserver projects?
Is there a reason Jamulus didn't just implement an existing protocol, and reuse an existing masterserver (there's even one being reimplemented in QT - https://code.333networks.com/333networks/MasterServer-Qt5)?

For masterserver purposes Jamulus fits perfectly in to the concept of just another game.

What are the main constraints of the current directory servers? CPU? Bandwidth? Admin-time?

Masterservers typically manage a far greater number of servers in the list than Jamulus seems to.

Masterservers are typically split by geography, not by game or map (a parallel to Jamulus' genre). Isn't this a more logical approach for a program preoccupied with latency?

**Comment by dcorson-ticino-com** [comment]
I presume that these gaming systems didn't exist when Jamulus was developed.
The first Jamulus release was 2006 if I remember correctly.

**Comment by ann0see** [comment]
> Masterservers are typically split by geography, not by game or map

I think it’s just due to historical reasons. I’m fact, they’re all in central europe as far as I know. However, directories don’t necessary need to have an ultra low ping-time since you‘d probably jam on other servers.

What we could do would to split them more geographically. But that woud mean, we need a volunteer and trusted server admin in America, Asia,… quite sure we have people from all over the world but is it worth putting the server admin burden on them?

**Reply by rdica** [reply]
What would actually be gained by grouping geographically versus the current configuration?

**Reply by ann0see** [reply]
Nothing big. Just that if there's a power shortage or environmental disaster, Jamulus will still work.

**Reply by gene96817** [reply]
Perhaps we just need one or two mirrors for the directory server.

**Reply by QJKX** [reply]
The masterservers are split by geography - but the more important detail is that the directories are effectively split by geography too.

Geography roughly correlates to ping.

Defaulting a client to a directory of the servers with the best ping (geographically close)  is better.
* Finding low ping servers is faster.
* It's harder to miss low ping servers that are in a non-default directory (really important for establishing jamulus in a new geography with low activity)
* It's slower to find high ping servers
* higher ping servers get less wasted traffic

I don't think pure geographic lists are optimal, but I think they'd  improve the experience for most players, and are a very KISS solution.

Perhaps a way of trying it would be to add geographic directories to the existing list.

**Reply by dcorson-ticino-com** [reply]
You keep saying it yourself, the important thing is the ping time and the list is already sorted according to ping.
To help find people who may speak your language (the only reason I can think of why one might want to know the physical location) that info is already available if provided by the server admin (and almost all admins do).
Maybe your connect window is not wide enough for you to see the location info?

(I note that we could maybe enforce servers from one geographic area to be all on one directory (although I don't see a simple way to do that).  Then the genre would have to be listed in the server info.  Then we will have people complaining that they want to have the genres together as the are only interested in servers dedicated to their particular favorite genre.  As the ping time corresponds a little with the location we are getting two birds with one stone as it is presently presented.)

**Reply by ann0see** [reply]
Good point. I think you should read https://github.com/jamulussoftware/jamulus/discussions/1035 for the reasons behind the genre split.

**Reply by QJKX** [reply]
> You keep saying it yourself, the important thing is the ping time and the list is already sorted according to ping.

100%

> the only reason I can think of why one might want to know the physical location
> As the ping time corresponds a little with the location

I agree that creating elaborate filter controls isn't going to help most people most of the time, but having good defaults really makes a difference.

At the moment I can see servers from New York London and Taipei on the same list.

If I'm only interested in servers with a ping of <30ms filtering by a distance of <5,000km (because we know the speed of light) is a crude but useful optimization.

>Then we will have people complaining that they want to have the genres together as the are only interested in servers dedicated to their particular favorite genre. 

There's no reason there can't be a genre filter, or a genre-based directory too.

Those people are probably going to be happy with the tradeoff because it gives their servers more visibility (with players "local" to the server) which should grow their genre, providing more opportunities to jam.
